File: JavaScript/4_kyu/NextBiggerNumberWithTheSameDigits.js
function nextBigger(n) {
  let newNumber = -1;

  let arrNumber = n.toString().split("");

  for (
    let indexFullNumber = n.toString().length - 2;
    indexFullNumber >= 0;
    indexFullNumber--
  ) {
    let firstPart = arrNumber.slice(0, indexFullNumber);
    let secondPart = arrNumber.slice(indexFullNumber, arrNumber.length);

    for (
      let indexSecondPart = 0;
      indexSecondPart < secondPart.length;
      indexSecondPart++
    ) {
      let secondPartWithoutIndex = secondPart.slice(0);
      secondPartWithoutIndex = secondPartWithoutIndex.sort();

      let numberInTest = secondPartWithoutIndex[indexSecondPart];
      secondPartWithoutIndex.splice(indexSecondPart, 1);

      let cadidateArr = [
        ...firstPart,
        ...numberInTest,
        ...secondPartWithoutIndex,
      ];

      let candidateNumber = parseInt(cadidateArr.join(""));

      if (candidateNumber > n) {
        newNumber = candidateNumber;
        break;
      }
    }

    if (newNumber != -1) {
      break;
    }
  }

  return newNumber;
}

File: Go/4_kyu/NextBiggerNumberWithTheSameDigits.go
package main

import (
	"sort"
	"strconv"
	"strings"
)

func NextBigger(n int) int {
	var newNumber int = -1
	arrNumber := strings.Split(strconv.Itoa(n), "")

	for i := len(arrNumber) - 2; i >= 0; i-- {
		firstPart := make([]string, len(arrNumber[:i]))
		copy(firstPart, arrNumber[:i])

		secondPart := make([]string, len(arrNumber[i:]))
		copy(secondPart, arrNumber[i:])

		for j := 0; j < len(secondPart); j++ {
			secondPartWithoutIndex := make([]string, len(secondPart))
			copy(secondPartWithoutIndex, secondPart)

			sort.Strings(secondPartWithoutIndex)

			numberInTheMidle := secondPartWithoutIndex[j]

			cadidateArr := append(firstPart, numberInTheMidle)
			cadidateArr = append(cadidateArr, secondPartWithoutIndex[:j]...)
			cadidateArr = append(cadidateArr, secondPartWithoutIndex[j+1:]...)

			var candidateNumber, _ = strconv.Atoi(strings.Join(cadidateArr[:], ""))

			if candidateNumber > n {
				newNumber = candidateNumber
				break
			}
		}

		if newNumber != -1 {
			break
		}

	}

	return newNumber
}
